go/src/url_shortener/cmd/end_to_end_tester/main.go

```diff
@@ -91,92 +137,171 @@ func testStatisticsJSON() error {
+		fmt.Println("expected ok response on redirect weather URL hash, but got:", response.StatusCode)
+		return err
+	}
+
+	requestURL := response.Request.URL
+	url := requestURL.String()
+
+	if url != weatherURLRome {
+		fmt.Println("expected redirect to", weatherURLRome, ", but got:", url)
+		return err
+	}
+
+	return nil
+}
 
-	fmt.Println("Starting test...")
+func testURLAddedToPersistenceFile() error {
+	urlShortener := shorten.NewURLShortener()
+
+	f, err := os.Open(persistenceFile)
+	if err != nil {
+		log.Println("error unpersisting URL added to persistence file:", err)
+		return err
+	}
+	defer f.Close()
+
+	reader := bufio.NewReader(f)
+	if err := urlShortener.UnpersistFrom(reader); err != nil {
+		fmt.Println("unable to unpersist URL added to persistence file:", persistenceFile, "err:", err)
+		return err
+	}
+
+	gotLongURL, err := urlShortener.GetURL(shortWeatherURLRome)
+	if err != nil {
+		fmt.Println("unable to get URL added to persistence file:", err)
+		return err
+	}
+
+	if weatherURLRome != gotLongURL {
+		fmt.Println("unable to find URL added to persistence file:", err)
+		return err
+	}
+
+	return nil
+}
+
+func trueMain(serverClosed chan struct{}) int {
+	cmd := exec.Command(serverExecutable, loadFlag, persistenceFile)
+	err := cmd.Start()
+
+	if err != nil {
+		log.Println(err)
+		return exitCodeError
+	}
+
+	go func(serverClosed chan struct{}) {
+		cmd.Wait()
+		close(serverClosed)
+	}(serverClosed)
+
+	time.Sleep(oneSecond) // to let the goroutine go
+
+	defer func() {
+		cmd.Process.Signal(syscall.SIGINT)
+		time.Sleep(oneSecond) // to let the kill kill
+	}()
 
 	if err := testNonExistentHash(); err != nil {
-		return 1
+		return exitCodeError
 	}
 
 	if err := testWeatherHash(); err != nil {
-		return 1
+		return exitCodeError
 	}
 
 	if err := testStatisticsJSON(); err != nil {
-		return 1
+		return exitCodeError
 	}
 
-	fmt.Println("Exit")
+	if err := testShortenURLAddingANonExistentURL(); err != nil {
+		return exitCodeError
+	}
 
-	return 0
-}
+	if err := testRedirectURL(); err != nil {
+		return exitCodeError
+	}
 
-/*
+	return exitCodeOk
+}
 
-Scenario:
+func main() {
+	serverClosed := make(chan struct{})
 
-1. call / on an non-existing SHA (check the http.StatusCode)
-2. call / on a existing SHA (loaded from the persistence file at startup)
-3. call /shorten with a new URL (that wasn't in the persistence file)
-4. call / with the SHA of the URL that has just been added
-        (check status code and ensure the redirect is working)
-5. call /statistics, unmarshall the json and checks that it corresponds to the actions taken
-6. terminate the http server
-7. verifies the content of the persistence file contains the URL added in step 3)
+	if err := copy(sourcePersistenceFile, persistenceFile); err != nil {
+		log.Println("persistence file not copied. Error:", err)
+		os.Exit(exitCodeError)
+	}
 
-*/
+	exitCode := trueMain(serverClosed)
 
-func main() {
-	exitCode := trueMain()
+	<-serverClosed
```

Comment:
Generally in Go we try to hide, whenever possible of course, the internals and make function calls synchronous.

I think trueMain should be blocking here, I'm not sure why the serverClosed channel has to be passed to that function, since just after the call we're doing a blocking receive on that channel.
Is there a reason I'm not seeing for that design? can't we just make trueMain synchronous and not have to deal with that channel (apart from inside trueMain of course)? 
trueMain

Reply:
The technical reason is the call to the `testURLAddedToPersistenceFile` function. This call has to be done when the server is closed (for obvious reasons), but the server is closed in a defer block inside `trueMain` which is called at `trueMain` exit. If you know a way to make the code working with the call inside `trueMain` I will be glad to hear that.

Reply:
Just put the cmd.Wait in the defer, after the kill and the sleep, like:

```diff
diff --git a/go/src/url_shortener/cmd/end_to_end_tester/main.go b/go/src/url_shortener/cmd/end_to_end_tester/main.go
index 1f87777..5cbba71 100644
--- a/go/src/url_shortener/cmd/end_to_end_tester/main.go
+++ b/go/src/url_shortener/cmd/end_to_end_tester/main.go
@@ -244,7 +244,7 @@ func testURLAddedToPersistenceFile() error {
 	return nil
 }
 
-func trueMain(serverClosed chan struct{}) int {
+func trueMain() int {
 	cmd := exec.Command(serverExecutable, loadFlag, persistenceFile)
 	err := cmd.Start()
 
@@ -253,16 +253,12 @@ func trueMain(serverClosed chan struct{}) int {
 		return exitCodeError
 	}
 
-	go func(serverClosed chan struct{}) {
-		cmd.Wait()
-		close(serverClosed)
-	}(serverClosed)
-
 	time.Sleep(oneSecond) // to let the goroutine go
 
 	defer func() {
 		cmd.Process.Signal(syscall.SIGINT)
 		time.Sleep(oneSecond) // to let the kill kill
+		cmd.Wait()
 	}()
 
 	if err := testNonExistentHash(); err != nil {
@@ -289,16 +285,13 @@ func trueMain(serverClosed chan struct{}) int {
 }
 
 func main() {
-	serverClosed := make(chan struct{})
-
 	if err := copy(sourcePersistenceFile, persistenceFile); err != nil {
 		log.Println("persistence file not copied. Error:", err)
 		os.Exit(exitCodeError)
 	}
 
-	exitCode := trueMain(serverClosed)
+	exitCode := trueMain()
 
-	<-serverClosed
 	if err := testURLAddedToPersistenceFile(); err != nil {
 		exitCode = exitCodeError
 	}

```

Happy to speak with you about if you want

Reply:
Sometimes the simplest solution is the hardest to see! :angel: